src/kubernetes/api_watcher.rs

```diff
@@ -152,3 +164,457 @@ pub mod invocation {
+                        match error {
+                            watcher::stream::Error::Desync { .. } => {}
+                            _ => panic!("expected a desync error"),
+                        }
+                    }),
+                    Box::new(|item| {
+                        assert_watch_response(
+                            item,
+                            WatchEvent::Added(Pod {
+                                metadata: ObjectMeta {
+                                    uid: Some("uid0".to_owned()),
+                                    ..Default::default()
+                                },
+                                ..Default::default()
+                            }),
+                        );
+                    }),
+                ],
+            ),
+            // Non-desync Stream Error
+            (
+                Box::new(|when, then| {
+                    when.method(GET).path("/api/v1/pods");
+                    then.status(200)
+                        .header("Content-Type", "application/json")
+                        .body(
+                            r#"{
+                            "type": "ERROR",
+                            "object": {
+                                "apiVersion": "v1",
+                                "code": 500,
+                                "kind": "Status",
+                                "message": "Internal Server Error",
+                                "metadata": {},
+                                "reason": "Puter go BOOM",
+                                "status": "Failure"
+                            }
+                        }"#,
+                        );
+                }),
+                vec![Box::new(|item| {
+                    assert_watch_response(
+                        item,
+                        WatchEvent::ErrorStatus(Status {
+                            code: Some(500),
+                            message: Some("Internal Server Error".to_owned()),
+                            reason: Some("Puter go BOOM".to_owned()),
+                            status: Some("Failure".to_owned()),
+                            ..Default::default()
+                        }),
+                    );
+                })],
+            ),
+            // Bad JSON from API
+            (
+                Box::new(|when, then| {
+                    when.method(GET).path("/api/v1/pods");
+                    then.status(200)
+                        .header("Content-Type", "application/json")
+                        .body(
+                            r#"
+                            "type": "ADDED",
+                            "object":
+                                "kind": "Pod",
+                                "apiVersion": "v1",
+                                "metadata": {
+                                    "uid": "uid0"
+                                }
+                            }
+                         }"#,
```

Comment:
Can we make it clearer this is an invalid JSON?
```suggestion
                            r#"not a valid json"#,
```